lib/Runtime/Library/JavascriptGenerator.cpp

```diff
@@ -247,6 +262,419 @@ namespace Js
+        {
+            int nextOffset = this->frame->GetReader()->GetCurrentOffset();
+            int endOffset = this->frame->GetFunctionBody()->GetByteCode()->GetLength();
+
+            if (nextOffset != endOffset - 1)
+            {
+                return;
+            }
+        }
+
+        ProcessAsyncGeneratorReturn(result, scriptContext);
+    }
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitYield(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwaitYield without a parameter");
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+        JavascriptGenerator* generator = resumeNextReturnProcessor->GetGenerator();
+        generator->SetState(GeneratorState::Suspended);
+
+        generator->AsyncGeneratorResolve(args[1], false);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitYieldStar(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwaitYield without a parameter");
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+        JavascriptGenerator* generator = resumeNextReturnProcessor->GetGenerator();
+        generator->SetState(GeneratorState::Suspended);
+
+        if (VarIs<RecyclableObject>(args[1]))
+        {
+            RecyclableObject* yieldData = UnsafeVarTo<RecyclableObject>(args[1]);
+            Var value = JavascriptOperators::GetProperty(yieldData, PropertyIds::value, scriptContext);
+
+            JavascriptOperators::OP_AsyncYield(generator, value, scriptContext);
+            return scriptContext->GetLibrary()->GetUndefined();
+        }
+        Var error = generator->CreateTypeError(JSERR_NonObjectFromIterable, scriptContext, _u("yield*"));
+        JavascriptOperators::OP_AsyncYield(generator, error, scriptContext);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    Var JavascriptGenerator::AsyncGeneratorEnqueue(Var thisValue, ScriptContext* scriptContext, Var input, JavascriptExceptionObject* exceptionObj, const char16* apiNameForErrorMessage)
+    {
+        JavascriptPromise* promise = JavascriptPromise::CreateEnginePromise(scriptContext);
+
+        if (!VarIs<JavascriptGenerator>(thisValue))
```

Comment:
Is it possible to hit this path without an internal error of some kind, and is it a case where the spec calls for a type error?

Reply:
Yes this would be if you did something like:
```js
async function* asyncGenF() {}
const nextMethod = asyncGenF().next;
nextMethod.call(anything);
```
Relevant spec text is here - it does have a type error for this:https://tc39.github.io/ecma262/#sec-asyncgeneratorenqueue

I'll add the spec text as comments to this function - not sure why I didn't add them before.